### PR TITLE
JDK-8258603 c1 IR::verify is expensive

### PR DESCRIPTION
**Pull request by LudwikJaniuk**
IR::verify iterates the whole object graph. This proves costly when used in e.g. BlockMerger inside of iterations over BlockLists, leading to quadratic or worse complexities as a function of bytecode length. In several cases, only a few Blocks were changed, and there was no need to go over the whole graph, but until now there was no less blunt tool for verification than IR::verify.

This PR introduces IR::verify_local, intended to be used when only a defined set of blocks have been modified. As a complement, expand_with_neighbors provides a way to also capture the neighbors of the "modified set" ahead of modification, so that afterwards the appropriate asserts can be made on all blocks which might possibly have been changed. All this should let us remove the expensive IR::verify calls, while still performing equivalent (or stricter) assertions.

Some changes have been made in the verifiers along the way. Some amount of refactoring, and even added invariants (see validate_edge_mutiality).

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8258603](https://bugs.openjdk.java.net/browse/JDK-8258603): c1 IR::verify is expensive

### Reviewers
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6850/head:pull/6850` \
`$ git checkout pull/6850`

Update a local copy of the PR: \
`$ git checkout pull/6850` \
`$ git pull https://git.openjdk.java.net/jdk pull/6850/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6850`

View PR using the GUI difftool: \
`$ git pr show -t 6850`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6850.diff">https://git.openjdk.java.net/jdk/pull/6850.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/c1/c1_IR.cpp [hunk 1]
```diff
@@ -1359,24 +1370,95 @@ class PredecessorValidator : public BlockClosure {
...
+  blocks.iterate_forward(&verifier);
+#endif // ASSERT
+}
+
 void IR::verify() {
 #ifdef ASSERT
-  PredecessorValidator pv(this);
-  EndNotNullValidator(this);
+  XentryFlagValidator xe;
+  this->iterate_postorder(&xe);
```
Comment (chhagedorn) [inline src/hotspot/share/c1/c1_IR.cpp:1449]: You can remove `this->`.

**Review by chhagedorn — CHANGES_REQUESTED** [review]
I think it's a good idea to narrow the verification down to the blocks which are actually changed and also add some additional verification steps. Even with the newly added checks, it should still be better than running a complete check each time. Did you observe a big gain in the C1 compilation time for debug builds with your new patch? Maybe you also want to check again with the tests from https://github.com/openjdk/jdk16/pull/44.

**Comment by LudwikJaniuk** [comment]
@chhagedorn Thank you for the review! I will get to cleaning up the comments, and I'll also get back about changes in compilations times.

**Comment by LudwikJaniuk** [comment]
Note that the timeouts of [openjdk/jdk16#44](https://github.com/openjdk/jdk16/pull/44) were remediated, likely by accident, in https://bugs.openjdk.java.net/browse/JDK-8267806, thus making it a worse candidate for testing this.

**Comment by LudwikJaniuk** [comment]
The increase in the previous test was likely not statistically significant, I ran a new test with 100 iterations and the difference seems gone.

100-2 iterations | C1 compile time |   | Emit LIR |   | Linear Scan |  
-- | -- | -- | -- | -- | -- | --
  | my fix | master | my fix | master | my fix | master
avg | 1,6151 | 1,617 | 1,3228 | 1,3238 | 1,2886 | 1,2893
variance | 0,0024 | 0,002 | 0,0022 | 0,0019 | 0,0021 | 0,0018

**Comment by LudwikJaniuk** [comment]
A test on -version shows a clear reduction in Optimize time, leading to a small reduction in c1 compile time. I call this a "boosted test" because `-XX:RepeatCompilation=10  -Xcomp -XX:TieredStopAtLevel=1` is used to make the time spent compiling in c1 longer and thereby measureable.
100-2 iterations | C1 comp time |   | Bulid HIR |   | Optimize |  
-- | -- | -- | -- | -- | -- | --
boosted -version | this PR | master | this PR | master | this PR | master
diff | −3% |   | −6% |   | −66,7% |  
avg | 1,6185 | 1,6685 | 0,7651 | 0,8143 | 0,0325 | 0,0976
variance | 0,0019 | 0,0019 | 0,0004 | 0,0005 | 0 | 0

Invocation:
` for i in {1..100}; do  .../jdk-master/bin/java  -XX:RepeatCompilation=10  -Xcomp -XX:+CITime -XX:TieredStopAtLevel=1  -version  > time/mast-$i.txt;  .../jdk-myfix/bin/java  -XX:RepeatCompilation=10  -Xcomp -XX:+CITime -XX:TieredStopAtLevel=1  -version > time/myfix-$i.txt; echo $i; done`

**Comment by LudwikJaniuk** [comment]
And again, it's important to remember that the point of this fix is to protect against "worst cases" with e.g. functions with very long bytecode. The compile time improvement in the general case, if any, is only a bonus.

File: src/hotspot/share/c1/c1_Optimizer.cpp [hunk 2]
```diff
@@ -387,6 +399,21 @@ class BlockMerger: public BlockClosure {
...
+#ifdef ASSERT
+    {
+      BlockList b;
+      b.append(block);
+      b.append(sux);
+      _hir->verify_local(b);
+    }
+#endif // ASSERT
+
+#ifdef ASSERT
```
[the remark this answers is not in the document]
Comment (LudwikJaniuk) [reply src/hotspot/share/c1/c1_Optimizer.cpp:411]: I'm removing that specific verify_local call anyway, left in by accident

**Comment by LudwikJaniuk** [comment]
@chhagedorn I believe I have addressed all your comments now.

**Review by chhagedorn — APPROVED** [review]
Thanks for doing the updates and pulling the numbers together! 

> And again, it's important to remember that the point of this fix is to protect against "worst cases" with e.g. functions with very long bytecode. The compile time improvement in the general case, if any, is only a bonus.

Yes, exactly. I think as long as are not noticeably worse on average (which we seem not to be according to your evaluation), we're good to go since this only affects debug builds.

Maybe someone else can also comment on the `ASSERT` vs. `NOT_PRODUCT` usage. I think we should either convert everything to one or the other but not mix it since all the verification code belongs together. Thinking about it again, `ASSERT` might be preferable to not affect the optimized build (as `IR::verify()` was guarded by `ASSERT` before).

**Comment by LudwikJaniuk** [comment]
I'll wait for the final word on ASSERT vs PRODUCT

File: src/hotspot/share/c1/c1_IR.cpp [hunk 3]
```diff
@@ -1414,24 +1398,23 @@ void IR::expand_with_neighborhood(BlockList& blocks) {
     BlockBegin* block = blocks.at(h);
 
     for (int i = 0; i < block->end()->number_of_sux(); i++) {
-      if (blocks.contains(block->end()->sux_at(i))) continue;
+      if (blocks.contains(block->end()->sux_at(i))) { continue; }
```
[the remark this answers is not in the document]
Comment (LudwikJaniuk) [reply src/hotspot/share/c1/c1_IR.cpp]: I went ahead and flipped the ifs instead

File: src/hotspot/share/c1/c1_Optimizer.cpp [hunk 4]
```diff
@@ -450,9 +437,9 @@ class BlockMerger: public BlockClosure {
                     _merge_count, block->block_id(), sux->block_id(), sux->state()->stack_size());
     }
 
-#ifdef ASSERT
+#ifndef PRODUCT
     _hir->verify_local(blocks_to_verify_later);
```
[the remark this answers is not in the document]
Comment (LudwikJaniuk) [reply src/hotspot/share/c1/c1_Optimizer.cpp]: Went with DEBUG_ONLY

File: src/hotspot/share/c1/c1_Optimizer.cpp [hunk 5]
```diff
@@ -180,6 +180,15 @@ void CE_Eliminator::block_do(BlockBegin* block) {
     return;
   }
 
+#ifdef ASSERT
```
[the remark this answers is not in the document]
Comment (vnkozlov) [reply src/hotspot/share/c1/c1_Optimizer.cpp:183]: No, `#ifdef ASSERT` is good here. `blocks_to_verify_later` was processed before by `IR::verify()` only in debug build (method's code was under `#ifdef ASSERT`). Now it passed to `IR::verify_local()` which is called only debug build.

**Review by vnkozlov — COMMENTED** [review]
So the whole block of code in `c1_IR.cpp` is under `#ifndef PRODUCT`.
But is it used in **optimized** build or only in debug?

I suggest to change `#ifndef PRODUCT` at line 1224 to `#ifdef ASSERT` and try to build optimized VM to see if it is used in it.
I see only prints and asserts.

**Comment by albertnetymk** [comment]
I wonder if "map fusion" works here, as in improving the perf of verification.

Using `IR::verify_local` as an example:

```
{
  VerifyClosure cl;
  blocks.iterate_forward(&cl);
}

class VerifyClosure : public BlockClosure {

  void block_do(BlockBegin* block) override {
    verify_end_not_null(block);
    
    verify_edge_mutuality(block);
    
    verify_block_begin_field(block);
    
    // more verifier
  }
}
```

This should cut down #iteration to 1.

**Comment by LudwikJaniuk** [comment]
@vnkozlov I don't know what an optimized build is, you'll have to point me to some instructions if you want me to build one. 

I think changing assertions around BlockPrinter is outside the scope of this PR. Perhaps another PR to clean up optimized builds is in order for the future?

**Comment by LudwikJaniuk** [comment]
@albertnetymk It would work for almost all of the verifiers, but there is a tradeoff with code flexibility. I'd opt to leave as-is.

**Comment by chhagedorn** [comment]
> So the whole block of code in `c1_IR.cpp` is under `#ifndef PRODUCT`. But is it used in **optimized** build or only in debug?
> 
> I suggest to change `#ifndef PRODUCT` at line 1224 to `#ifdef ASSERT` and try to build optimized VM to see if it is used in it. I see only prints and asserts.

Shouldn't we leave `BlockPrinter` and `IR::print()` under `#ifndef PRODUCT`? These seem to be about printing things only which is probably good to have in optimized builds as well. We could instead just add an `#ifdef ASSERT` on L1263 for the validators and the IR verification and change the places using this code accordingly to `#ifdef ASSERT/DEBUG_ONLY()`. Then we would have everything under `ASSERT` instead of `!PRODUCT`. This would be in line with the original code which used `#ifdef ASSERT` in `IR::verify()`.

**Comment by vnkozlov** [comment]
> > So the whole block of code in `c1_IR.cpp` is under `#ifndef PRODUCT`. But is it used in **optimized** build or only in debug?
> > I suggest to change `#ifndef PRODUCT` at line 1224 to `#ifdef ASSERT` and try to build optimized VM to see if it is used in it. I see only prints and asserts.
> 
> Shouldn't we leave `BlockPrinter` and `IR::print()` under `#ifndef PRODUCT`? These seem to be about printing things only which is probably good to have in optimized builds as well. We could instead just add an `#ifdef ASSERT` on L1263 for the validators and the IR verification and change the places using this code accordingly to `#ifdef ASSERT/DEBUG_ONLY()`. Then we would have everything under `ASSERT` instead of `!PRODUCT`. This would be in line with the original code which used `#ifdef ASSERT` in `IR::verify()`.

I agree with christian's suggestion.

**Comment by vnkozlov** [comment]
> @vnkozlov I don't know what an optimized build is, you'll have to point me to some instructions if you want me to build one.

```
bash configure --with-conf-name=optimized --with-debug-level=optimized
make jdk-image CONF=optimized
```

**Comment by LudwikJaniuk** [comment]
```
./build/optimized/images/jdk/bin/java -XX:+PrintCFG -version
CFG after parsing
B1 [0, 0] -> B2
B2 (S) [0, 0] -> B0 pred: B1
B0 (SV) [0, 0] pred: B2

CFG after optimizations
B1 [0, 0] -> B2
B2 (S) [0, 0] -> B0 pred: B1
B0 (SV) [0, 0] pred: B2

CFG after null check elimination
B1 [0, 0] -> B2
B2 (S) [0, 0] -> B0 dom B1 pred: B1
B0 (SV) [0, 0] dom B2 pred: B2

CFG before code generation
B1 [0, 0] -> B2
B2 (S) [0, 0] -> B0 dom B1 pred: B1
B0 (SV) [0, 0] dom B2 pred: B2
openjdk version "19-internal" 2022-09-20
OpenJDK Runtime Environment (build 19-internal+0-adhoc.opjaniuk.jdk)
OpenJDK 64-Bit Server VM (build 19-internal+0-adhoc.opjaniuk.jdk, mixed mode, sharing)
```
Is this as it should be?

**Comment by LudwikJaniuk** [comment]
@vnkozlov You proposed changing c1_IR.cpp:1244 (I think you meant :1224) to `#ifdef ASSERT`. I tried it with an optimized build and this gave linking errors, specifically "undefined reference to IR::verify()." Seems likely to me that this is caused by mismatch withuse of `PRODUCT_RETURN` in c1_IR.hpp.

**Comment by LudwikJaniuk** [comment]
Whichever way the PRODUCT/ASSERT pendulum swings, what's for sure is that it needs to be consistent, so I decided to move it around "single source of truth"-style.

I'm not from the compiler team so I don't know the implications or the build process, but I'm surprised there isn't an answer to  "how do we want to guard debug code?" I need a clear decision and guideline from the compiler team here.

**Comment by chhagedorn** [comment]
> @vnkozlov You proposed changing c1_IR.cpp:1244 (I think you meant :1224) to `#ifdef ASSERT`. I tried it with an optimized build and this gave linking errors, specifically "undefined reference to IR::verify()." Seems likely to me that this is caused by mismatch withuse of `PRODUCT_RETURN` in c1_IR.hpp.

When applying these changes:

> > > So the whole block of code in `c1_IR.cpp` is under `#ifndef PRODUCT`. But is it used in **optimized** build or only in debug?
> > > I suggest to change `#ifndef PRODUCT` at line 1224 to `#ifdef ASSERT` and try to build optimized VM to see if it is used in it. I see only prints and asserts.
> > 
> > 
> > Shouldn't we leave `BlockPrinter` and `IR::print()` under `#ifndef PRODUCT`? These seem to be about printing things only which is probably good to have in optimized builds as well. We could instead just add an `#ifdef ASSERT` on L1263 for the validators and the IR verification and change the places using this code accordingly to `#ifdef ASSERT/DEBUG_ONLY()`. Then we would have everything under `ASSERT` instead of `!PRODUCT`. This would be in line with the original code which used `#ifdef ASSERT` in `IR::verify()`.
> 
> I agree with christian's suggestion.

Then you can change `PRODUCT_RETURN` into `NOT_DEBUG_RETURN` to compile it and change `__DO_DELAYED_VERIFICATION` back to `ASSERT`.

My understanding is that assertion/verification code should be under `ASSERT` while additional debugging/printing code that is not available in product should be under `!PRODUCT`. But I don't know if that is documented somewhere officially and there are some places where it is mixed. I think there were also some discussions to move away from optimized builds completely.

**Comment by chhagedorn** [comment]
> ```
> ./build/optimized/images/jdk/bin/java -XX:+PrintCFG -version
> CFG after parsing
> B1 [0, 0] -> B2
> B2 (S) [0, 0] -> B0 pred: B1
> B0 (SV) [0, 0] pred: B2
> 
> CFG after optimizations
> B1 [0, 0] -> B2
> B2 (S) [0, 0] -> B0 pred: B1
> B0 (SV) [0, 0] pred: B2
> 
> CFG after null check elimination
> B1 [0, 0] -> B2
> B2 (S) [0, 0] -> B0 dom B1 pred: B1
> B0 (SV) [0, 0] dom B2 pred: B2
> 
> CFG before code generation
> B1 [0, 0] -> B2
> B2 (S) [0, 0] -> B0 dom B1 pred: B1
> B0 (SV) [0, 0] dom B2 pred: B2
> openjdk version "19-internal" 2022-09-20
> OpenJDK Runtime Environment (build 19-internal+0-adhoc.opjaniuk.jdk)
> OpenJDK 64-Bit Server VM (build 19-internal+0-adhoc.opjaniuk.jdk, mixed mode, sharing)
> ```
> 
> Is this as it should be?

Yes, that looks good. It accepts the non-product flag `PrintCFG`

**Comment by vnkozlov** [comment]
> @vnkozlov You proposed changing c1_IR.cpp:1244 (I think you meant :1224) to `#ifdef ASSERT`. I tried it with an optimized build and this gave linking errors, specifically "undefined reference to IR::verify()." Seems likely to me that this is caused by mismatch withuse of `PRODUCT_RETURN` in c1_IR.hpp.

I did say 1224: "I suggest to change #ifndef PRODUCT at line 1224 to #ifdef ASSERT"

As Christian said, use `NOT_DEBUG_RETURN` for cases when the method body is defined only in debug build (under `#ifdef ASSERT`).

Note, my suggestion for changing `!PRODUCT` to `ASSERT` at line 1224 and build `optimized` VM was experiment we frequently use to find which functions are not use in `optimized` VM and which are used. It is usually faster than tracing usage in sources. I don't mean to use it in final changes. By using this experiment you can separate methods between `!PRODUCT` and `ASSERT`.

Please,  use `#ifdef ASSERT` or `DEBUG_ONLY()` instead of `__DO_DELAYED_VERIFICATION`. 

> My understanding is that assertion/verification code should be under ASSERT while additional debugging/printing code that is not available in product should be under !PRODUCT. But I don't know if that is documented somewhere officially and there are some places where it is mixed. I think there were also some discussions to move away from optimized builds completely.

Yes, this is correct. The assertion/verification is used in our regular testing when we use fastdebug builds to catch issues.
Optimized build is mostly used to understand what is going on in build which is very similar to product.

Yes, we have RFE to remove optimized build and replace it with separate value check:
https://bugs.openjdk.java.net/browse/JDK-8202283

**Comment by chhagedorn** [comment]
> Yes, this is correct. The assertion/verification is used in our regular testing when we use fastdebug builds to catch issues. Optimized build is mostly used to understand what is going on in build which is very similar to product.

Thanks for confirming that.

> Yes, we have RFE to remove optimized build and replace it with separate value check: https://bugs.openjdk.java.net/browse/JDK-8202283

That's good to know!

**Comment by LudwikJaniuk** [comment]
There might be a prettier alternative than `__DO_DELAYED_VERIFICATION`, but I want to stress that my delayed-verification mechanism needs to be enabled in full, or not at all. I worry someone might change from ASSERT to !PRODUCT or vice versa on one but not all of the parts in the future. I worry that might lead to hard-to-debug problems. So, are you sure you'd like me to drop this SSOT-approach? If you're sure, I'll do it the way you wish.

So I understand the conclusion was to use ASSERT, NOT_DEBUG, and NOT_DEBUG_RETURN for all the code in this PR. Thank you.

@vnkozlov apologies if I got the line numbering wrong in my head somewhere there.

**Comment by vnkozlov** [comment]
I am not comfortable with repeated `#ifdef` around verification methods calls. But you have to pass local `blocks_to_verify_later` which is defined only in debug VM.

I'm fine with using `__DO_DELAYED_VERIFICATION` but add comment to the definition explaining what it is for.
Also don't use leading `__` - we don't do that.

I think you can simply put verification methods declaration undo `ASSERT` in `c1_IR.hpp`:
```
   void print(bool cfg_only, bool live_only = false)                           PRODUCT_RETURN;
-  void verify()                                                               PRODUCT_RETURN;
+
+#ifdef ASSERT
+  void verify();
+  void expand_with_neighborhood();
+  void verify_local();
+#endif // ASSERT
 };
```

And put everything from `class EndNotNullValidator` to `IR::verify()` under `#ifdef ASSERT` in `c1_IR.cpp`.
(And move `#endif // PRODUCT` before `class EndNotNullValidator`.)
This way verification code will be defined only in debug VM and print code will be in optimized and debug VM.

**Comment by vnkozlov** [comment]
And please update copyright year to 2022 in changed files headers.

**Comment by LudwikJaniuk** [comment]
@vnkozlov Done, although using ASSERT in c1_IR.hpp didn't work because c1_Compilation.cpp makes several unguarded calls to IR::verify. I've left it with NOT_DEBUG_RETURN, hope that is ok.

I hope I've covered all the issues now, but please point it out if I've missed something.

**Review by chhagedorn — APPROVED** [review]
> @vnkozlov Done, although using ASSERT in c1_IR.hpp didn't work because c1_Compilation.cpp makes several unguarded calls to IR::verify. I've left it with NOT_DEBUG_RETURN, hope that is ok.

I think that is fine. Otherwise, you need to wrap the calls in `c1_compilation.cpp` with `NOT_DEBUG()`.

Thanks for doing the updates, they look good to me! If not already done, please verify once again that the optimized build is working with your latest changes before integration ;-)

**Comment by LudwikJaniuk** [comment]
Yes, I did verify that.

**Review by vnkozlov — APPROVED** [review]
Looks good.

**Comment by LudwikJaniuk** [comment]
@vnkozlov @chhagedorn Could you please sponsor?

**Pull request closed without merging**